tslearn/metrics.py

```diff
@@ -207,6 +209,188 @@ def dtw_path(s1, s2, global_constraint=None, sakoe_chiba_radius=None,
     return path, numpy.sqrt(acc_cost_mat[-1, -1])
 
 
+@njit()
+def njit_accumulated_matrix_from_dist_matrix(dist_matrix, mask):
+    """Compute the accumulated cost matrix score between two time series using
+    a precomputed distance matrix.
+
+    Parameters
+    ----------
+    dist_matrix : array, shape = (sz1, sz2)
+        Array containing the pairwise distances.
+
+    mask : array, shape = (sz1, sz2)
+        Mask. Unconsidered cells must have infinite values.
+
+    Returns
+    -------
+    mat : array, shape = (sz1, sz2)
+        Accumulated cost matrix.
+
+    """
+    l1, l2 = dist_matrix.shape
+    cum_sum = numpy.full((l1 + 1, l2 + 1), numpy.inf)
+    cum_sum[0, 0] = 0.
+
+    for i in prange(l1):
+        for j in prange(l2):
+            if numpy.isfinite(mask[i, j]):
+                cum_sum[i + 1, j + 1] = dist_matrix[i, j]
+                cum_sum[i + 1, j + 1] += min(cum_sum[i, j + 1],
+                                             cum_sum[i + 1, j],
+                                             cum_sum[i, j])
+    return cum_sum[1:, 1:]
+
+
+def dtw_path_from_metric(s1, s2=None, metric="precomputed",
+                         global_constraint=None, sakoe_chiba_radius=None,
+                         itakura_max_slope=None, **kwds):
+    r"""Compute Dynamic Time Warping (DTW) similarity measure between
+    (possibly multidimensional) time series using a distance metric defined by
+    the user and return both the path and the similarity.
+
+    Similarity is computed as the cumulative cost along the wrapped time
+    series.
+
+    It is not required that both time series share the same size, but they must
+    be the same dimension. DTW was originally presented in [1]_.
+
+    Valid values for metric are the same as for scikit-learn
+    `pairwise_distances`_ function I.E. a string (E.G. "euclidean",
+    "sqeuclidean", "hamming") or a function that is used to compute the
+    pairwise distances. See `scikit`_ and `scipy`_ documentations for more
+    information about the available metrics.
+
+    Parameters
+    ----------
+    s1 : array, shape = (sz1, sz2) if metric=="precomputed", (sz1,) otherwise
```

Comment:
I'd rather set the `(sz1, d)` case as default than the precomputed case. (also add the feature dimension d for s2)

Reply:
Works for me, I'll also change the default metric to "euclidean" instead of "precomputed" to be consistent with the default input.